File: acgs2-core/sdk/typescript/src/services/audit.ts
/**
 * ACGS-2 Audit Service
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import { ACGS2Client } from '../client';
import {
  CONSTITUTIONAL_HASH,
  AuditEvent,
  AuditEventSchema,
  EventCategory,
  EventSeverity,
  QueryAuditEventsRequest,
  PaginatedResponse,
} from '../types';
import { ValidationError, nowISO, generateUUID } from '../utils';

// =============================================================================
// Audit Types
// =============================================================================

export interface CreateAuditEventRequest {
  category: EventCategory;
  severity: EventSeverity;
  action: string;
  actor: string;
  resource: string;
  resourceId?: string;
  outcome: 'success' | 'failure' | 'partial';
  details?: Record<string, unknown>;
  correlationId?: string;
}

export interface AuditTrail {
  id: string;
  resourceType: string;
  resourceId: string;
  events: AuditEvent[];
  firstEvent: string;
  lastEvent: string;
  constitutionalHash: string;
}

export interface AuditExport {
  id: string;
  format: 'json' | 'csv' | 'parquet';
  status: 'pending' | 'processing' | 'completed' | 'failed';
  query: QueryAuditEventsRequest;
  createdAt: string;
  completedAt?: string;
  downloadUrl?: string;
  recordCount?: number;
  constitutionalHash: string;
}

export interface AuditStatistics {
  totalEvents: number;
  byCategory: Record<string, number>;
  bySeverity: Record<string, number>;
  byOutcome: Record<string, number>;
  topActors: Array<{ actor: string; count: number }>;
  topResources: Array<{ resource: string; count: number }>;
  eventsOverTime: Array<{ timestamp: string; count: number }>;
  constitutionalHash: string;
}

// =============================================================================
// Audit Service
// =============================================================================

export class AuditService {
  private readonly basePath = '/api/v1/audit';

  constructor(private readonly client: ACGS2Client) {}

  // ===========================================================================
  // Event Management
  // ===========================================================================

  /**
   * Records an audit event
   */
  async record(event: CreateAuditEventRequest): Promise<AuditEvent> {
    const response = await this.client.post<AuditEvent>(`${this.basePath}/events`, {
      ...event,
      timestamp: nowISO(),
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to record audit event');
    }

    return this.validateEvent(response.data);
  }

  /**
   * Records multiple audit events in batch
   */
  async recordBatch(events: CreateAuditEventRequest[]): Promise<AuditEvent[]> {
    const preparedEvents = events.map((event) => ({
      ...event,
      timestamp: nowISO(),
      constitutionalHash: CONSTITUTIONAL_HASH,
    }));

    const response = await this.client.post<AuditEvent[]>(
      `${this.basePath}/events/batch`,
      { events: preparedEvents }
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to record audit events batch');
    }

    return response.data.map((e) => this.validateEvent(e));
  }

  /**
   * Gets an audit event by ID
   */
  async getEvent(eventId: string): Promise<AuditEvent> {
    const response = await this.client.get<AuditEvent>(`${this.basePath}/events/${eventId}`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Audit event not found: ${eventId}`);
    }

    return this.validateEvent(response.data);
  }

  /**
   * Queries audit events with filters
   */
  async queryEvents(params: QueryAuditEventsRequest): Promise<PaginatedResponse<AuditEvent>> {
    const response = await this.client.get<PaginatedResponse<AuditEvent>>(
      `${this.basePath}/events`,
      params as Record<string, unknown>
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to query audit events');
    }

    response.data.data = response.data.data.map((e) => this.validateEvent(e));
    return response.data;
  }

  /**
   * Searches audit events with full-text search
   */
  async searchEvents(
    query: string,
    params?: QueryAuditEventsRequest
  ): Promise<PaginatedResponse<AuditEvent>> {
    const response = await this.client.get<PaginatedResponse<AuditEvent>>(
      `${this.basePath}/events/search`,
      { query, ...params } as Record<string, unknown>
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to search audit events');
    }

    response.data.data = response.data.data.map((e) => this.validateEvent(e));
    return response.data;
  }

  // ===========================================================================
  // Audit Trails
  // ===========================================================================

  /**
   * Gets the audit trail for a resource
   */
  async getTrail(resourceType: string, resourceId: string): Promise<AuditTrail> {
    const response = await this.client.get<AuditTrail>(
      `${this.basePath}/trails/${resourceType}/${resourceId}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Audit trail not found: ${resourceType}/${resourceId}`);
    }

    return response.data;
  }

  /**
   * Gets audit trails for multiple resources
   */
  async getTrailsBatch(
    resources: Array<{ resourceType: string; resourceId: string }>
  ): Promise<AuditTrail[]> {
    const response = await this.client.post<AuditTrail[]>(`${this.basePath}/trails/batch`, {
      resources,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get audit trails batch');
    }

    return response.data;
  }

  // ===========================================================================
  // Export
  // ===========================================================================

  /**
   * Creates an export job for audit events
   */
  async createExport(options: {
    format: 'json' | 'csv' | 'parquet';
    query?: QueryAuditEventsRequest;
  }): Promise<AuditExport> {
    const response = await this.client.post<AuditExport>(`${this.basePath}/exports`, {
      ...options,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to create audit export');
    }

    return response.data;
  }

  /**
   * Gets an export job status
   */
  async getExport(exportId: string): Promise<AuditExport> {
    const response = await this.client.get<AuditExport>(`${this.basePath}/exports/${exportId}`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Audit export not found: ${exportId}`);
    }

    return response.data;
  }

  /**
   * Lists export jobs
   */
  async listExports(): Promise<AuditExport[]> {
    const response = await this.client.get<AuditExport[]>(`${this.basePath}/exports`);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list audit exports');
    }

    return response.data;
  }

  /**
   * Waits for an export to complete
   */
  async waitForExport(
    exportId: string,
    options?: { timeoutMs?: number; pollIntervalMs?: number }
  ): Promise<AuditExport> {
    const timeout = options?.timeoutMs ?? 300000; // 5 minutes default
    const pollInterval = options?.pollIntervalMs ?? 5000; // 5 seconds default
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const exportJob = await this.getExport(exportId);

      if (exportJob.status === 'completed') {
        return exportJob;
      }

      if (exportJob.status === 'failed') {
        throw new ValidationError(`Export job failed: ${exportId}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new ValidationError(`Export job timed out: ${exportId}`);
  }

  // ===========================================================================
  // Statistics & Analytics
  // ===========================================================================

  /**
   * Gets audit statistics
   */
  async getStatistics(params?: {
    startDate?: string;
    endDate?: string;
    category?: EventCategory;
  }): Promise<AuditStatistics> {
    const response = await this.client.get<AuditStatistics>(
      `${this.basePath}/statistics`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get audit statistics');
    }

    return response.data;
  }

  /**
   * Gets activity summary for an actor
   */
  async getActorActivity(
    actor: string,
    params?: { startDate?: string; endDate?: string }
  ): Promise<{
    actor: string;
    totalEvents: number;
    byCategory: Record<string, number>;
    byOutcome: Record<string, number>;
    recentEvents: AuditEvent[];
    constitutionalHash: string;
  }> {
    const response = await this.client.get<{
      actor: string;
      totalEvents: number;
      byCategory: Record<string, number>;
      byOutcome: Record<string, number>;
      recentEvents: AuditEvent[];
      constitutionalHash: string;
    }>(`${this.basePath}/actors/${encodeURIComponent(actor)}/activity`, params);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to get actor activity: ${actor}`);
    }

    return response.data;
  }

  /**
   * Gets activity summary for a resource
   */
  async getResourceActivity(
    resource: string,
    resourceId: string,
    params?: { startDate?: string; endDate?: string }
  ): Promise<{
    resource: string;
    resourceId: string;
    totalEvents: number;
    byAction: Record<string, number>;
    byActor: Record<string, number>;
    timeline: AuditEvent[];
    constitutionalHash: string;
  }> {
    const response = await this.client.get<{
      resource: string;
      resourceId: string;
      totalEvents: number;
      byAction: Record<string, number>;
      byActor: Record<string, number>;
      timeline: AuditEvent[];
      constitutionalHash: string;
    }>(`${this.basePath}/resources/${resource}/${resourceId}/activity`, params);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to get resource activity: ${resource}/${resourceId}`);
    }

    return response.data;
  }

  // ===========================================================================
  // Retention & Archival
  // ===========================================================================

  /**
   * Gets retention policy
   */
  async getRetentionPolicy(): Promise<{
    defaultRetentionDays: number;
    byCategory: Record<string, number>;
    archiveEnabled: boolean;
    archiveDestination?: string;
    constitutionalHash: string;
  }> {
    const response = await this.client.get<{
      defaultRetentionDays: number;
      byCategory: Record<string, number>;
      archiveEnabled: boolean;
      archiveDestination?: string;
      constitutionalHash: string;
    }>(`${this.basePath}/retention`);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get retention policy');
    }

    return response.data;
  }

  // ===========================================================================
  // Integrity Verification
  // ===========================================================================

  /**
   * Verifies the integrity of audit events
   */
  async verifyIntegrity(params?: {
    startDate?: string;
    endDate?: string;
    eventIds?: string[];
  }): Promise<{
    verified: boolean;
    totalChecked: number;
    validCount: number;
    invalidCount: number;
    invalidEvents: string[];
    constitutionalHash: string;
  }> {
    const response = await this.client.post<{
      verified: boolean;
      totalChecked: number;
      validCount: number;
      invalidCount: number;
      invalidEvents: string[];
      constitutionalHash: string;
    }>(`${this.basePath}/verify`, {
      ...params,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to verify audit integrity');
    }

    return response.data;
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  private validateEvent(data: unknown): AuditEvent {
    const result = AuditEventSchema.safeParse(data);
    if (!result.success) {
      throw new ValidationError('Invalid audit event data', {
        validation: result.error.errors.map((e) => e.message),
      });
    }
    return result.data;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/audit.py
"""
ACGS-2 Audit Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = str | int | float | bool | None | dict[str, Any] | list[Any]  # type: ignore[misc]
    JSONDict = dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import AUDIT_ENDPOINT, CONSTITUTIONAL_HASH
from acgs2_sdk.models import (
    AuditEvent,
    EventCategory,
    EventSeverity,
    PaginatedResponse,
    QueryAuditEventsRequest,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class AuditService:
    """Service for audit logging and querying."""

    def __init__(self, client: "ACGS2Client") -> None:
        self._client = client
        self._base_path = AUDIT_ENDPOINT

    async def record(
        self,
        category: EventCategory,
        severity: EventSeverity,
        action: str,
        actor: str,
        resource: str,
        outcome: str,
        resource_id: str | None = None,
        details: JSONDict | None = None,
        correlation_id: str | None = None,
    ) -> AuditEvent:
        """Record an audit event."""
        data = await self._client.post(
            f"{self._base_path}/events",
            json={
                "category": category.value,
                "severity": severity.value,
                "action": action,
                "actor": actor,
                "resource": resource,
                "resourceId": resource_id,
                "outcome": outcome,
                "details": details,
                "correlationId": correlation_id,
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return AuditEvent.model_validate(data.get("data", data))

    async def record_batch(
        self,
        events: list[JSONDict],
    ) -> list[AuditEvent]:
        """Record multiple audit events."""
        prepared = [
            {
                **event,
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            }
            for event in events
        ]
        data = await self._client.post(
            f"{self._base_path}/events/batch",
            json={"events": prepared},
        )
        return [AuditEvent.model_validate(e) for e in data.get("data", [])]

    async def get_event(self, event_id: str) -> AuditEvent:
        """Get an audit event by ID."""
        data = await self._client.get(f"{self._base_path}/events/{event_id}")
        return AuditEvent.model_validate(data.get("data", data))

    async def query_events(
        self,
        request: QueryAuditEventsRequest | None = None,
        **kwargs: Any,
    ) -> PaginatedResponse[AuditEvent]:
        """Query audit events."""
        params = request.model_dump(by_alias=True, exclude_none=True) if request else kwargs

        data = await self._client.get(f"{self._base_path}/events", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[AuditEvent](
            data=[AuditEvent.model_validate(e) for e in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", 1),
            page_size=response_data.get("pageSize", 50),
            total_pages=response_data.get("totalPages", 0),
        )

    async def search_events(
        self,
        query: str,
        page: int = 1,
        page_size: int = 50,
    ) -> PaginatedResponse[AuditEvent]:
        """Search audit events."""
        data = await self._client.get(
            f"{self._base_path}/events/search",
            params={"query": query, "page": page, "pageSize": page_size},
        )
        response_data = data.get("data", data)
        return PaginatedResponse[AuditEvent](
            data=[AuditEvent.model_validate(e) for e in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def get_trail(
        self,
        resource_type: str,
        resource_id: str,
    ) -> JSONDict:
        """Get audit trail for a resource."""
        data = await self._client.get(f"{self._base_path}/trails/{resource_type}/{resource_id}")
        return data.get("data", data)

    async def create_export(
        self,
        format: str = "json",
        query: QueryAuditEventsRequest | None = None,
    ) -> JSONDict:
        """Create an export job."""
        payload: JSONDict = {
            "format": format,
            "constitutionalHash": CONSTITUTIONAL_HASH,
        }
        if query:
            payload["query"] = query.model_dump(by_alias=True, exclude_none=True)

        data = await self._client.post(f"{self._base_path}/exports", json=payload)
        return data.get("data", data)

    async def get_export(self, export_id: str) -> JSONDict:
        """Get export job status."""
        data = await self._client.get(f"{self._base_path}/exports/{export_id}")
        return data.get("data", data)

    async def get_statistics(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
        category: EventCategory | None = None,
    ) -> JSONDict:
        """Get audit statistics."""
        params: JSONDict = {}
        if start_date:
            params["startDate"] = start_date
        if end_date:
            params["endDate"] = end_date
        if category:
            params["category"] = category.value

        data = await self._client.get(f"{self._base_path}/statistics", params=params)
        return data.get("data", data)

    async def verify_integrity(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
        event_ids: list[str] | None = None,
    ) -> JSONDict:
        """Verify audit event integrity."""
        data = await self._client.post(
            f"{self._base_path}/verify",
            json={
                "startDate": start_date,
                "endDate": end_date,
                "eventIds": event_ids,
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return data.get("data", data)
